iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java

```diff
@@ -133,42 +145,43 @@ public IotHubTransportMessage receive()
+                        this.receivedMessages.poll();
+
+                        //parse the topic string to get information (method name, request Id, etc.) further
+                        TopicParser topicParser = new TopicParser(topic);
+
+                        if (data != null && data.length > 0)
+                        {
+                            message = new IotHubTransportMessage(data, MessageType.DEVICE_METHODS);
+                        }
+                        else
                         {
-                            //remove this message from the queue as this is the correct handler
-                            this.receivedMessages.poll();
-
-                            // Case for $iothub/methods/POST/{method name}/?$rid={request id}
-                            TopicParser topicParser = new TopicParser(topic);
-
-                            if (data != null && data.length > 0)
-                            {
-                                message = new IotHubTransportMessage(data, MessageType.DEVICE_METHODS);
-                            }
-                            else
-                            {
-                                message = new IotHubTransportMessage(new byte[0], MessageType.DEVICE_METHODS);
-                            }
-
-                            message.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_UNKNOWN);
-                            message.setQualityOfService(mqttMessage.getQos());
-
-                            String methodName = topicParser.getMethodName(METHOD_TOKEN);
-                            message.setMethodName(methodName);
-
-                            String reqId = topicParser.getRequestId(REQID_TOKEN);
-                            if (reqId != null)
-                            {
-                                message.setRequestId(reqId);
-
-                                message.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_METHOD_RECEIVE_REQUEST);
-                            }
-                            else
-                            {
-                                log.warn("Request ID cannot be null");
-                            }
+                            message = new IotHubTransportMessage(new byte[0], MessageType.DEVICE_METHODS);
+                        }
+
+                        message.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_UNKNOWN);
+                        message.setQualityOfService(mqttMessage.getQos());
+
+                        String methodName = this.isConnectingToMqttGateway ?
+                                topicParser.getMethodName(4) :
+                                topicParser.getMethodName(3);
+                        message.setMethodName(methodName);
+
+                        String reqId = this.isConnectingToMqttGateway ?
+                                topicParser.getRequestId(5) :
+                                topicParser.getRequestId(4);
```

Comment:
Similar here, getting requestId from the field 5 if connecting to E4K, or 4 if not.

($iothub/methods/testDevice/POST/testMethod/?$**rid=10**)
($iothub/methods/POST/testMethod/?**$rid=10**)